[Qt] Add option to pause/resume block downloads

jonasschnelli opened the pull request:
This, almost UI only change, will add a Pause/Resume button to the modal overlay to pause/resume block downloads during IBD.

This is an effective way to pause/resume IBD during a time when the computers resources are required somewhere else.

<img width="962" alt="bildschirmfoto 2017-01-10 um 18 34 09" src="https://cloud.githubusercontent.com/assets/178464/21817212/659f314c-d763-11e6-87e2-5b0754d067e0.png">

jonasschnelli committed:
[Qt] update header-syncing progress when not processing a block

jonasschnelli committed:
Add fAutoRequestBlocks to disabled/enable the verification progress

jonasschnelli committed:
[Qt] Add option to pause/resume block downloads

paveljanik commented:
Isn't network toggle button usable in this case? If it is not, let's fix it instead...

jonasschnelli commented:
I extracted this from my SPV branch. Especially there, it is very useful to pause IBD and continue with SPV during a time where you don't want to use all available resources on verification.

But also without SPV, I think this can be useful (pause IBD and not loose broadcast capabilities, fetch headers but not the blocks)...

luke-jr commented:
IMO it would be too confusing to be worth it without "SPV" mode, but probably should go in after the latter is.

MarcoFalke commented:
> Isn't network toggle button usable in this case? If it is not, let's fix it instead...

Imo those are different features, but I agree that the GUI should not
"diverge" in regard to presenting features. The toggle network
functionality should be removed from the network icon and a proper
button should be put beside the new "Pause" button?

luke-jr commented:
The "Pause" button won't be visible in normal cases.

jonasschnelli commented:
>> Isn't network toggle button usable in this case? If it is not, let's fix it instead...

> Imo those are different features, but I agree that the GUI should not
"diverge" in regard to presenting features. The toggle network
functionality should be removed from the network icon and a proper
button should be put beside the new "Pause" button?

Yes. These are internally two completely different features. Expose to the users, these have similar effects.

> The "Pause" button won't be visible in normal cases.

Yes. The modal overlay is currently only accessible during IBD. Though, we could extend it to support a state where the chain is in-sync and show it when someone click on the network statusbar icon.

kallewoof commented on src/net_processing.h:
```diff
@@ -14,6 +14,10 @@ void RegisterNodeSignals(CNodeSignals& nodeSignals);
 /** Unregister a network node */
 void UnregisterNodeSignals(CNodeSignals& nodeSignals);
 
+/** if disabled, blocks will not be requested automatically, usefull for low-resources-available mode */
```
Typo: useful~l~ (one L)

kallewoof commented on src/qt/clientmodel.h:
```diff
@@ -81,6 +81,10 @@ class ClientModel : public QObject
     QString dataDir() const;
 
+    // get/set state about autorequesting-blocks during IBD
+    bool isAutorequestBlocks() const;
```
Nit: `isAutorequestingBlocks` (add `ing`). The `setAutorequestBlocks` below is fine as is, though.

kallewoof reviewed (commented):
utACK abf0941

laanwj commented:
Added 0.15 milestone

jonasschnelli commented:
Rebased.

laanwj commented on src/net_processing.h:
```diff
@@ -23,6 +23,10 @@ void RegisterNodeSignals(CNodeSignals& nodeSignals);
 
+/** if disabled, blocks will not be requested automatically, usefull for low-resources-available mode */
+static const bool DEFAULT_AUTOMATIC_BLOCK_REQUESTS = true;
+extern std::atomic<bool> fAutoRequestBlocks;
```
Instead of exporting this variable (whose definition is an implementation detail) from net_processing.h I'd prefer a setter/getter, e.g. `SetAutoRequestBlocks(bool)` `GetAutoRequestBlocks()`

jonasschnelli replied:
Agree. Will change...

laanwj commented on src/qt/clientmodel.cpp:
```diff
@@ -233,6 +234,22 @@ QString ClientModel::dataDir() const
     return GUIUtil::boostPathToQString(GetDataDir());
 }
 
+bool ClientModel::isAutorequestBlocks() const
```
isAutorequestBlocks is a bit of a strange name; isAutoRequestingBlocks maybe?

laanwj commented:
When re-enabling AutoRequestBlocks, what 'kicks off' the block requesting process again? Setting the flag will make it request blocks the next time `FindNextBlocksToDownload` is called; is that good enough? I suppose it is, because `SendMessages` is called periodically (every 100 ms?).

jonasschnelli committed:
Rename autorequest blocks methods

jonasschnelli committed:
Add setter/getter for fAutoRequestBlocks

jonasschnelli committed:
Show blocks-in-flight

jonasschnelli committed:
Show special info when autodownload is disabled but blocks are in flight

jonasschnelli commented:
Overhauled the PR and addresses @laanwj points.

Also, I added the info "Blocks requested from peers" (blocks in flight). This may be important because pause will not result in disconnecting peers. Already requested blocks will be downloaded (and verified) in the "pause" state.

If blocks are in flight and the pause has been triggered, there is now a special info label `Wait to finish current downloads`.

paveljanik commented:
Well, I have to change my previous opinion. I think this can be useful.

Concept ACK

Will test soon.

paveljanik commented:
Needs rebase.

ryanofsky commented:
ACK e952b67d294043d9758b37bf2be57a982b41c051

Code looks good. On the UI feature, two minor comments:

- Pausing the download throws off "Estimated time left till synced." It would be better pausing didn't affect estimated time.
- Pause/Resume button really sticks out where it's currently placed. Maybe it would make more sense next to the progress bar or near the hide button.

sipa commented:
Needs rebase.

jonasschnelli commented:
Rebased.

paveljanik commented:
Testing this again.

ACK https://github.com/bitcoin/bitcoin/pull/9502/commits/fc84323d8ee447f1461a10b7f3b29d113f9f4a43

I think this could be even more usable if it can be called once fully in sync with the network. But I can't display the overlay then...

laanwj commented:
> I think this could be even more usable if it can be called once fully in sync with the network. But I can't display the overlay then...

Hm I vaguely remember I added that functionality once, you should be able to bring up the overlay by the secret trick of clicking on the sync icon.

paveljanik commented:
@laanwj But the sync icon (you mean the triangle with an exclamation mark inside?) is not displayed when you are "in sync" with the network.

jonasschnelli commented:
> I think this could be even more usable if it can be called once fully in sync with the network. But I can't display the overlay then...

There are four ways how the modal-overlay can be opened:
-> auto-opens when in IBD/sync
-> Click on the warning icons next to the balance
-> Click on the progress bar during IBD/sync
-> Click on the sync icon

Though I agree with you, there is no option how to open it once you are in sync... which could be useful, but independent to this PR.

laanwj commented:
@paveljanik Oh, though it also worked with the checkmark that takes its place.

jonasschnelli commented:
> @paveljanik Oh, though it also worked with the checkmark that takes its place.

Yes. We should probably allow that.

sdaftuar commented:
Concept ACK. (I haven't reviewed the qt code at all, but I did look over the net_processing changes.)

What is the intended behavior if the user turns off block download after being synced or nearly synced?  I believe this PR only disables the parallel fetch logic, and not the direct fetch -- so block download could conceivably be disabled by the user, but blocks could still be requested as they are announced.  (Maybe this is hard to accomplish, if the modal overlay is not accessible when bitcoind is close to synced, but my recollection is that the modal overlay becomes visible whenever the headers chain is out of sync, which can happen even after leaving IBD, does that sound right?)

Perhaps the ability to disable block download should disappear if we're close to caught up, eg if `CanDirectFetch()` is true?

Additionally, perhaps if we've disabled block download, we should also disable transaction download.

jonasschnelli committed:
Disable "pause downloading blocks" when we can fetch blocks directly

jonasschnelli commented:
Rebased and addressed the `CanDirectFetch()` issue @sdaftuar mentioned.

The current implementation avoids exposing `net_processing.cpp`'s `CanDirectFetch()` for the reasons a) it does always use chainActive (lock) and b) to avoid another core-layer dependency.

I'm not doing to tackle the transaction download in the PR (main scope is disabling block download).

achow101 commented:
ACK

I have tested this on a node that was a few weeks behind and I was able to pause the block download multiple times while it was syncing. I noticed that even when paused, UpdateTip is still happening for hundreds of blocks after what the GUI said still needed to be downloaded. I assume this is because it still processes all of the blocks that have been downloaded and the downloading happens faster than the processing. The UpdateTip lines do stop after a while which I assume is because all downloaded blocks have been processed.

sipa commented:
> I assume this is because it still processes all of the blocks that have been downloaded and the downloading happens faster than the processing.

Yes, that's expected. We download multiple blocks in parallel, ahead of the next one to connect. I expect that this PR also keeps processing the ones that were in flight at the time blocks downloads are disabled.

promag commented on src/qt/forms/modaloverlay.ui:
```diff
@@ -334,6 +354,33 @@ QLabel { color: rgb(40,40,40);  }</string>
             <number>10</number>
            </property>
            <item>
+            <widget class="QPushButton" name="pauseResumeVerification">
```
Nit, rename `toggleDownloadButton`.

jonasschnelli replied:
I prefer to keep it because it's not pure download, it's the whole verification (and maybe we further extend what it does in the background later).

promag commented on src/qt/forms/modaloverlay.ui:
```diff
@@ -334,6 +354,33 @@ QLabel { color: rgb(40,40,40);  }</string>
             <number>10</number>
            </property>
            <item>
```
Nit, can't comment there, but remove `leftMargin` above to be consistent.

jonasschnelli replied:
Why remove?

promag commented on src/qt/modaloverlay.cpp:
```diff
@@ -170,3 +189,21 @@ void ModalOverlay::closeClicked()
+{
+    ui->labelNumberBlocksRequested->setText((pauseActive ? "Finish downloading blocks": "Blocks requested from peers"));
+    ui->pauseResumeVerification->setText((pauseActive ? "Resume downloading blocks ": "Pause downloading blocks"));
+    ui->infoLabel->setText((pauseActive && getAmountOfBlocksInFlight() > 0 ? "Wait to finish current downloads...": ""));
```
I don't know if you plan to show other messages for this label, but if not then toggle visibility like above?

jonasschnelli replied:
I think this is fine how it is, as long as `nBlocksInFlight > 0` we need to expect network traffic and therefor should still remind the user with the `"Wait to finish current..."`.

promag commented on src/qt/modaloverlay.cpp:
```diff
@@ -170,3 +189,21 @@ void ModalOverlay::closeClicked()
+{
+    ui->labelNumberBlocksRequested->setText((pauseActive ? "Finish downloading blocks": "Blocks requested from peers"));
+    ui->pauseResumeVerification->setText((pauseActive ? "Resume downloading blocks ": "Pause downloading blocks"));
+    ui->infoLabel->setText((pauseActive && getAmountOfBlocksInFlight() > 0 ? "Wait to finish current downloads...": ""));
```
Also, when paused, `nBlocksInFlight` doesn't always decrease to zero.

promag commented on src/qt/modaloverlay.cpp:
```diff
@@ -170,3 +189,21 @@ void ModalOverlay::closeClicked()
+
+void ModalOverlay::updatePauseState(bool pauseActive)
+{
+    ui->labelNumberBlocksRequested->setText((pauseActive ? "Finish downloading blocks": "Blocks requested from peers"));
```
Don't see the need to change label, maybe a generic `Downloading: %d blocks`?

jonasschnelli replied:
I think the currents PR label switch makes it more clear what happens in the background (pause syncing will not lead to immediate network download halt).

laanwj commented:
This has missed the 0.15 feature freeze, moving to 0.16.

jonasschnelli commented:
Rebased.

jonasschnelli commented:
rebased and slightly overhauled the blocks-in-flight information flow.

jonasschnelli commented:
Rebased.
@paveljanik @achow101 @ryanofsky @kallewoof: care to re-ack?

kallewoof commented on src/net_processing.cpp:
```diff
@@ -3306,6 +3312,19 @@ bool PeerLogicValidation::SendMessages(CNode* pto, std::atomic<bool>& interruptM
+    return fAutoRequestBlocks;
+}
+
+unsigned int getAmountOfBlocksInFlight() {
```
I think they should all start with capital or non-capital (`SetAuto` vs `isAuto`/`getAmount`) for consistency.

MeshCollider replied:
Also nit, the braces should be on new lines

kallewoof reviewed (approved):
utACK 1375063565a257057707379588b02a8aa44b7bcc (my QT skills are weak though)

paveljanik commented on src/net_processing.h:
```diff
@@ -61,4 +64,10 @@ bool GetNodeStateStats(NodeId nodeid, CNodeStateStats &stats);
+void SetAutoRequestBlocks(bool);
+bool isAutoRequestingBlocks();
+
+/** retruns the amount of blocks in flight (in total) */
```
typo retruns

paveljanik commented on src/qt/modaloverlay.cpp:
```diff
@@ -127,15 +131,30 @@ void ModalOverlay::tipUpdate(int count, const QDateTime& blockDate, double nVeri
+    eventuallyShowHeaderSyncing(count);
+    updatePauseState(verificationPauseActive);
+
+    // disable pause button when we we can fetch directly
```
we we

MeshCollider commented on src/qt/modaloverlay.cpp:
```diff
@@ -127,15 +131,30 @@ void ModalOverlay::tipUpdate(int count, const QDateTime& blockDate, double nVeri
+
+    // disable pause button when we we can fetch directly
+    // avoid using the core-layer's existing CanFetchDirectly()
+    bool canFetchDirecly = (blockDate.toTime_t() > GetAdjustedTime() - Params().GetConsensus().nPowTargetSpacing * 20);
```
Typo in variable name `canFetchDirecly`, should be `Directly`

MeshCollider commented on src/qt/modaloverlay.cpp:
```diff
@@ -172,3 +191,21 @@ void ModalOverlay::closeClicked()
+
+void ModalOverlay::updatePauseState(bool pauseActive)
+{
+    ui->labelNumberBlocksRequested->setText((pauseActive ? "Finish downloading blocks": "Blocks requested from peers"));
```
nit, space missing before `:` (same on next two lines)

MeshCollider commented on src/qt/modaloverlay.cpp:
```diff
@@ -172,3 +191,21 @@ void ModalOverlay::closeClicked()
+void ModalOverlay::updatePauseState(bool pauseActive)
+{
+    ui->labelNumberBlocksRequested->setText((pauseActive ? "Finish downloading blocks": "Blocks requested from peers"));
+    ui->pauseResumeVerification->setText((pauseActive ? "Resume downloading blocks ": "Pause downloading blocks"));
```
nit, space at the end of the first string instead of before `:`

meshcollider reviewed (commented):
utACK https://github.com/bitcoin/bitcoin/pull/9502/commits/1375063565a257057707379588b02a8aa44b7bcc

Just a couple of formatting nits inline

jonasschnelli commented:
Rebased and addressed nits (code style).

kallewoof commented on src/net_processing.cpp:
```diff
@@ -3647,6 +3653,21 @@ bool PeerLogicValidation::SendMessages(CNode* pto, std::atomic<bool>& interruptM
+    return fAutoRequestBlocks;
+}
+
+unsigned int getAmountOfBlocksInFlight() {
```
Brace starts on same line for this function but on different lines for above two functions. I think separate line is the preferred style these days, but I could be wrong?

kallewoof reviewed (commented):
utACK f820e8645071e66b33fd97bcaaab8516ca429f77

jonasschnelli commented:
@cfields: since this is touching net code, can you do a final review?

fanquake commented:
Had a quick test of this (b0a9b8d6dae2a9efcdbc3607b235c37e1f7db31d) on top of master (9b8b1079ddab64ac955766536c38d23dc57bc499). macOS 10.13.3 and Qt 5.10.1.

This is what the overlay model looks like now with this PR (b0a9b8d6dae2a9efcdbc3607b235c37e1f7db31d):
![147511](https://user-images.githubusercontent.com/863730/37718009-b363ad0e-2d5c-11e8-9ff8-d01169983d10.png)

One thing I noticed is after hitting "Pause", it can take a while to stop downloading blocks. 
i.e In the images below, we hit pause with 127 blocks "requested from peers" and 145022 total blocks left. However we don't appear to stop downloading blocks until we've reached 144601 total blocks left to download, > 400 blocks later. This could be confusing, and might look worse on slower connections?

<img width="1119" alt="145022" src="https://user-images.githubusercontent.com/863730/37718535-e1ba009e-2d5d-11e8-95c7-89fc5c418a21.png">
<img width="1133" alt="144601" src="https://user-images.githubusercontent.com/863730/37718555-ec53ed94-2d5d-11e8-96d9-1933eec59f75.png">

I'll comment inline about simplifying the new strings.

fanquake commented on src/qt/modaloverlay.cpp:
```diff
@@ -172,3 +191,21 @@ void ModalOverlay::closeClicked()
+
+void ModalOverlay::updatePauseState(bool pauseActive)
+{
+    ui->labelNumberBlocksRequested->setText((pauseActive ? "Finish downloading blocks" : "Blocks requested from peers"));
```
I'd suggest changing this string to just "Blocks in download queue" or similar. Then we don't have to worry about switching strings between paused and downloading.

fanquake commented on src/qt/modaloverlay.cpp:
```diff
@@ -172,3 +191,21 @@ void ModalOverlay::closeClicked()
+{
+    ui->labelNumberBlocksRequested->setText((pauseActive ? "Finish downloading blocks" : "Blocks requested from peers"));
+    ui->pauseResumeVerification->setText((pauseActive ? "Resume downloading blocks " : "Pause downloading blocks"));
+    ui->infoLabel->setText((pauseActive && getAmountOfBlocksInFlight() > 0 ? "Wait to finish current downloads..." : ""));
```
Maybe use "Downloading blocks remaining in queue.." or similar here?

jonasschnelli commented:
> One thing I noticed is after hitting "Pause", it can take a while to stop downloading blocks.
i.e In the images below, we hit pause with 127 blocks "requested from peers" and 145022 total blocks left. However we don't appear to stop downloading blocks until we've reached 144601 total blocks left to download, > 400 blocks later. This could be confusing, and might appear worse on slower connections?

Yes. That could be confusing,.. though I guess there is an explanation. You may still have blocks on your disk that hasn't been verified due to a missing the next blocks after your tip. This means it is possible to jump a couple of block after downloading a single block.

Also,... we don't want to ignore the request because the peer is sending it anyway (once requested) and we don't want to disconnect peers (that feature is already available).

It would be possible to hide that information from the user, though I think it makes more sense to display it so its clear that the "pause" means not "disconnect".

jonasschnelli commented:
Rebased.
Thanks for re-reviewing.

MarcoFalke commented on src/qt/bitcoingui.cpp:
```diff
@@ -493,6 +493,10 @@ void BitcoinGUI::setClientModel(ClientModel *_clientModel)
         modalOverlay->setKnownBestHeight(_clientModel->getHeaderTipHeight(), QDateTime::fromTime_t(_clientModel->getHeaderTipTime()));
+        modalOverlay->setPauseResumeState(!_clientModel->isAutoRequestingBlocks());
+        connect(modalOverlay, SIGNAL(requestVerificationPauseOrResume()), _clientModel, SLOT(toggleAutoRequestBlocks()));
+        connect(_clientModel, SIGNAL(verificationProgressPauseStateHasChanged(bool)), modalOverlay, SLOT(setPauseResumeState(bool)));
```
could use the new connect syntax, so it wouldn't have to be changed again in the future?

jonasschnelli commented:
Rebased.

practicalswift commented on src/qt/modaloverlay.cpp:
```diff
@@ -127,15 +131,30 @@ void ModalOverlay::tipUpdate(int count, const QDateTime& blockDate, double nVeri
+
+    // disable pause button when we can fetch directly
+    // avoid using the core-layer's existing CanFetchDirectly()
+    bool canFetchDirectly = (blockDate.toTime_t() > GetAdjustedTime() - Params().GetConsensus().nPowTargetSpacing * 20);
```
nit: Redundant parentheses :-)

fanquake commented:
Discussion has dragged on and died out here. Going to label "Up for grabs" and close for now.
